neptune/internal/api_clients/hosted_api_clients/migration_switch_leaderboard_api_client_proxy.py

```diff
@@ -0,0 +1,52 @@
+#
+# Copyright (c) 2021, Neptune Labs Sp. z o.o.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import logging
+import threading
+
+from neptune.backend import LeaderboardApiClient
+
+from neptune.exceptions import ProjectMigratedToNewStructure
+from neptune.internal.api_clients import HostedNeptuneBackendApiClient
+from neptune.internal.api_clients.hosted_api_clients.hosted_leaderboard_api_client import \
+    HostedNeptuneLeaderboardApiClient
+
+_logger = logging.getLogger(__name__)
+
+
+# pylint: disable=abstract-method
+class MigrationSwitchLeaderboardApiClientProxy(LeaderboardApiClient):
+
+    def __init__(self, api_client: HostedNeptuneLeaderboardApiClient, backend_client: HostedNeptuneBackendApiClient):
+        self._client = api_client
+        self._backend_client = backend_client
+
+        self._lock = threading.RLock()
+        self._switched = False
+
+    def __getattr__(self, item):
+        def func(*args, **kwargs):
+            try:
+                return getattr(self._client, item)(*args, **kwargs)
+            except ProjectMigratedToNewStructure:
+                if not self._switched:
```

Comment:
I like this - this will allow us to change endpoints live witch running clients , but this self._switched bothers me - we can do this only once per client run cycle.

Can you remove it? I am aware that this will cause more GET-s for swagger

Reply:
No, we can not get rid of it. It is important for threads synchronization. All threads (std, metrics, channels, main thread and ping) use the same backend object and it is highly probable that they will get this exception at the same time.